packages/react/src/hooks/useVisibilityObserver.test.js

```diff
@@ -0,0 +1,141 @@
+
+    expect(onChange).toHaveBeenCalledWith({ isVisible: true });
+    rerender(initialProps);
+    expect(result.current).toEqual([true]);
+
+    // unobserve should be called once because the element is now visible, and should not be observed
+    // anymore since unobserveAfterVisible is true
+    expect(observer.unobserve).toHaveBeenCalledTimes(1);
+    expect(observer.disconnect).toHaveBeenCalledTimes(1);
+  });
+
+  it('should continue to observe all changes when unobserveAfterVisible:false', () => {
+    const initialProps = [ref, { unobserveAfterVisible: false, onChange }];
+    const { result } = renderHook(
+      ([elementRef, options]) => useVisibilityObserver(elementRef, options),
+      {
+        initialProps,
+      }
+    );
+
+    // the initial render should observe the ref one time
+    expect(observer.observe).toHaveBeenCalledWith(ref.current);
+    expect(observer.observe).toHaveBeenCalledTimes(1);
+    expect(result.current).toEqual([false]);
+    expect(onChange).not.toHaveBeenCalled();
+
+    // mock the observer calling the callback when the element is visible
+    act(() => {
+      isIntersecting = true;
+      observerCallback([{ isIntersecting }], observer);
+    });
+
+    expect(observer.observe).toHaveBeenCalledTimes(1);
+    expect(observer.unobserve).not.toHaveBeenCalled();
+    expect(onChange).toHaveBeenCalledWith({ isVisible: true });
+    expect(result.current).toEqual([true]);
+
+    // unobserve should not be called because the element is now visible, but we're still observing it
+    expect(observer.unobserve).not.toHaveBeenCalled();
+    expect(observer.disconnect).not.toHaveBeenCalled();
+
+    // mock the observer calling the callback when the element is hidden again
+    act(() => {
+      isIntersecting = false;
+      observerCallback([{ isIntersecting }], observer);
+    });
+
+    expect(observer.unobserve).not.toHaveBeenCalled();
+    expect(observer.disconnect).not.toHaveBeenCalled();
+    expect(onChange).toHaveBeenCalledWith({ isVisible: false });
+    expect(result.current).toEqual([false]);
+  });
+
+  it('should disconnect the existing and create a new observer when the ref changes', () => {
+    const initialProps = [ref, { onChange }];
+    const { result, rerender } = renderHook(
+      ([elementRef, options]) => useVisibilityObserver(elementRef, options),
+      {
+        initialProps,
+      }
+    );
+
+    // the initial render should observe the ref one time
+    expect(observer.observe).toHaveBeenCalledWith(ref.current);
+    expect(observer.observe).toHaveBeenCalledTimes(1);
+    expect(result.current).toEqual([false]);
+    expect(onChange).not.toHaveBeenCalled();
+
+    const newRef = React.createRef();
+    newRef.current = document.createElement('div');
+    rerender([newRef, initialProps[1]]);
+
+    expect(observer.unobserve).not.toHaveBeenCalled();
+    expect(observer.observe).toHaveBeenCalledTimes(2);
+    expect(observer.disconnect).toHaveBeenCalledTimes(2);
```

Comment:
Why has `observer.disconnect` been called twice?

Reply:
once for changing to the new ref. I'm honestly not certain about the second. I tried to track it down for a long time. I can't figure out where it's coming from unless the useEffect cleanup is happening oddly in jest. I don't have any issues in storybook.